### ISSUE TITLE
nodejs: There is no way to unsubscribe from a stream

### ISSUE THREAD

**Issue by 123wowow123** [BitMEX/api-connectors#364]
For the nodejs connectors: api-connectors/official-ws/nodejs/

Once you subscribe to a stream via ` clientAuth.addStream("ETHXBT", 'order', orderListener)`
There is no way to unsubscribe from it

**Comment by mariusk** [comment]
Yes, they really should add endpoints for unsubscribing. In the meantime you may want to try:

`clientAuth.removeListener('order', orderListener);`

Also missing is the ability to close the automatically reconnecting websocket in an orderly manner (if you want your trading bot to exit / restart gracefully for instance). A workaround for this is:

`clientAuth.socket.instance.close(1000);`

where 1000 is an internal code used by the library for a "normal closing" of the socket (without triggering reconnect). If you don't close everything down properly, your Node.js process typically hangs around forever waiting for you to clean things up.

**Comment by 123wowow123** [comment]
Hi thanks for the response but executing `clientAuth.removeListener('order', orderListener);` still doesn't unregister the listener function. :(

**Comment by mariusk** [comment]
Yes, I think you're right. My use case was just cleaning up, and those calls didn't fail so I mistakenly assumed they worked. Did some more testing and I'm having more success with:

`clientAuth.removeAllListeners();`

At least I'm not getting any more callbacks with data. That still doesn't allow you to individually unsubscribe though.

There seems to be some support for stopping individual streams here:

https://github.com/BitMEX/api-connectors/blob/master/official-ws/nodejs/index.js#L161

.. but I'm not very versed in Node.js event emitters, so not sure how to invoke that code. Looks like it wants an event string like `order:SOMETHING:ETHXBT` to trigger. But again, not sure how to invoke that. Best of luck though.

**Comment by andersea** [comment]
I don't think it is easy to add, the way the code is designed. I looked at it seriously several times during the past couple of years that I have send patches occationally. There is just too much spaghetti code in there.

What I would recommend doing is to close the entire websocket and reopen a new one. This is what the sample-market-maker does, essentially. As long as you don't hit any websocket rate limit, this should be pretty straightforward.

I find that, in practice, I know ahead of time what channels I am going to need, so there isn't really any practical use in being able to unsubscribe.

**Comment by LucasRodriguez** [comment]
I figured this one out for y'all
as mentioned by @mariusk this should do the trick:

`clientAuth.socket.instance._events.close("1000")`

how ever you will also need to modify the file: 

> ReconnectingSocket.js

In the switch statement you will have to add an actual closure of the `ws` like so:

```
switch (code){
      case CLOSE_NORMAL:
        reconnecting = false;
        this.instance.close(code);
        this.instance.emit('close', code);
        this.logError(`WebSocket closed normally.`);
        break;
      case CLOSE_UNEXPECTED:
        break;
      default:    // Abnormal closure
        reconnecting = true;
        break;
   }
```

I will submit a PR for this change.

**Comment by LucasRodriguez** [comment]
Tested this by building a docker container with a call to create the ws and a call to close it. 
I then substituted the 

`ReconnectingSocket.js` with my own version through a volume. 

Meaning, I replaced the file after building node modules.

**Comment by guyappy** [comment]
@LucasRodriguez  I did find that after a while I get the following error with your code:
Uncaught RangeError: Maximum call stack size exceeded
get readyState           	websocket.js:173
close                    	websocket.js:269
(anonymous function)     	ReconnectingSocket.js:32

Line 32 would be " this.instance.close(code)" in this case. I haven't really figured out yet why it's doing this though. Maybe you have any thoughts?
```
  this.instance.on('close', (code) => {
    let reconnecting = false;

    switch (code){
      case CLOSE_NORMAL:
        reconnecting = false;
        this.instance.close(code);
        this.instance.emit('close', code);
        this.logError(`WebSocket closed normally.`);
        break;
      case CLOSE_UNEXPECTED:
        break;
      default:    // Abnormal closure
        reconnecting = true;
        break;
    }
```
